au/overflow_boundary_test.cc

```diff
@@ -0,0 +1,285 @@
+    // What all these test cases have in common is that the destination type is already the most
+    // constraining factor.  Therefore, the only way to add an _explicit_ limit, which nevertheless
+    // does _not_ constrain the answer, is to make that explicit limit equal to the implicit limit:
+    // that is, the lowest value of the destination type.
+
+    EXPECT_THAT((MinGood<StaticCast<int8_t, int8_t>, ImplicitLimits<int8_t>>::value()),
+                Eq(MinGood<StaticCast<int8_t, int8_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<uint16_t, uint16_t>, ImplicitLimits<uint16_t>>::value()),
+                Eq(MinGood<StaticCast<uint16_t, uint16_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<float, float>, ImplicitLimits<float>>::value()),
+                Eq(MinGood<StaticCast<float, float>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<uint32_t, int32_t>, ImplicitLimits<int32_t>>::value()),
+                Eq(MinGood<StaticCast<uint32_t, int32_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<int64_t, uint64_t>, ImplicitLimits<uint64_t>>::value()),
+                Eq(MinGood<StaticCast<int64_t, uint64_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<double, float>, ImplicitLimits<float>>::value()),
+                Eq(MinGood<StaticCast<double, float>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<float, uint64_t>, ImplicitLimits<uint64_t>>::value()),
+                Eq(MinGood<StaticCast<float, uint64_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<float, int64_t>, ImplicitLimits<int64_t>>::value()),
+                Eq(MinGood<StaticCast<float, int64_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<float, int32_t>, ImplicitLimits<int32_t>>::value()),
+                Eq(MinGood<StaticCast<float, int32_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<uint32_t, uint16_t>, ImplicitLimits<uint16_t>>::value()),
+                Eq(MinGood<StaticCast<uint32_t, uint16_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<uint32_t, int8_t>, ImplicitLimits<int8_t>>::value()),
+                Eq(MinGood<StaticCast<uint32_t, int8_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<int64_t, int32_t>, ImplicitLimits<int32_t>>::value()),
+                Eq(MinGood<StaticCast<int64_t, int32_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<int64_t, uint32_t>, ImplicitLimits<uint32_t>>::value()),
+                Eq(MinGood<StaticCast<int64_t, uint32_t>>::value()));
+}
+
+TEST(StaticCast, MinGoodUnchangedWithExplicitLimitLessConstrainingThanExistingResult) {
+    // In these cases, we are applying a non-trivial lower limit (i.e., it is higher than the
+    // `lowest()` value), but it does not constrain the result enough to change it.
+
+    struct DoubleLimitTwiceFloatLowest : NoUpperLimit<double> {
+        static constexpr double lower() {
+            return static_cast<double>(std::numeric_limits<float>::lowest()) * 2.0;
```

Comment:
In this case, multiplying by `2` as opposed to `2.0` would give the same result, correct?

```suggestion
            return static_cast<double>(std::numeric_limits<float>::lowest()) * 2;
```

Reply:
Yes, but I preferred to take an explicitly single-type approach, rather than mixing `double` and `int`.  This explains my approach in the other two situations you commented on below.  Is there a compelling reason to switch to the "int-shaped" versions, other than conciseness?